### PR TITLE
unused_parens now fires on cast expression

### LINKED ISSUE #10557: Unnecssary parenthesis lint
### What it does

Currently, if the code contains unnecessary parenthesis, clippy doesn't warn the user about it.  It would be nice if there was some new lint to detect these cases and warn the user about them. 

### Lint Name

_No response_

### Category

style

### Advantage

- Remove unnecessary parenthesis in the code

### Drawbacks

_No response_

### Example

```rust
fn new_rect(x: f64, y: f64, width: f64, height: f64) {
    println!("I am a rectangle ({x},{y}) ({width}, {height})");
}

fn main() {
    let x = 0f32;
    let y = 0f32;
    let width = 100f32;
    let height = 100f32;

    new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
}

```

Could be written as:

```rust
fn new_rect(x: f64, y: f64, width: f64, height: f64) {
    println!("I am a rectangle ({x},{y}) ({width}, {height})");
}

fn main() {
    let x = 0f32;
    let y = 0f32;
    let width = 100f32;
    let height = 100f32;

    new_rect(x as f64, y as f64, width as f64, height as f64);
}

```

### PR DESCRIPTION
**Pull request by KisaragiEffective**
close https://github.com/rust-lang/rust-clippy/issues/10557.
close https://github.com/rust-lang/rust-clippy/pull/10625.

@llogiq suggested submit PR to solve the issue.

### CODE REVIEW

File: tests/ui/lint/unused_parens_cast.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// check-pass
+
+#![warn(unused_parens)]
+
+fn hello(f: f64) {}
+
+fn main() {
+    hello((1.0f32) as f64) // ~WARN unnecessary parentheses
```
Comment (b-naber) [inline tests/ui/lint/unused_parens_cast.rs:8]: Can you maybe add more test cases here that test [this](https://github.com/rust-lang/rust/blob/9541afaa167846ec7bee04a9851346d964fc60c1/compiler/rustc_lint/src/unused.rs#L574-L584) more elabortately

**Review by b-naber — COMMENTED** [review]
Some nits, but LGTM.

Also you shouldn't have any changes in submodules here.

File: compiler/rustc_lint/src/unused.rs [hunk 2]
```diff
@@ -558,6 +560,7 @@ trait UnusedDelimLint {
         inner: &ast::Expr,
         followed_by_block: bool,
         followed_by_else: bool,
+        followed_by_cast: bool,
```
[the remark this answers is not in the document]
Comment (KisaragiEffective) [reply compiler/rustc_lint/src/unused.rs]: Sure, done!

File: tests/ui/lint/unused_parens_cast.rs (hunk 1 above)
Comment (KisaragiEffective) [reply tests/ui/lint/unused_parens_cast.rs]: added more testcases :)

**Comment by KisaragiEffective** [comment]
(force-pushed to drop accidental changes in stdarch submodule)

**Comment by KittyBorgX** [comment]
> should it be fixed in this PR as well as other occurrences?

@KisaragiEffective I'd suggest doing so if it isn't against the reviewer's descretion since that'll be the only way the CI passes and the compiler builds fully :)

**Comment by KisaragiEffective** [comment]
submitted https://github.com/rust-lang/stdarch/pull/1411, marking this as a draft until it is synced.

**Comment by KisaragiEffective** [comment]
It's merged 🚀 
Should I update the submodule in this PR or will they become updated eventually? :)

**Comment by KisaragiEffective** [comment]
> Can you please squash your commits?

Sure! However, I'd like to rebase (and squash) after stdarch update is landed in order to reduce force-push, so I'll submit another PR first.

**Comment by KisaragiEffective** [comment]
submitted #110285, this PR remains to be a draft.

**Comment by KisaragiEffective** [comment]
I've squahsed commits :)

**Comment by KisaragiEffective** [comment]
It looks like another PR to stdarch is needed.

File: src/tools/miri/tests/pass/float.rs [hunk 3]
```diff
@@ -198,7 +198,7 @@ fn casts() {
     assert_eq::<i32>(f32::INFINITY as i32, i32::MAX);
     assert_eq::<i32>(f32::NEG_INFINITY as i32, i32::MIN);
     assert_eq::<i32>(f32::NAN as i32, 0);
-    assert_eq::<i32>((-f32::NAN) as i32, 0);
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/float.rs:201]: I don't think we should lint against `(-expr) as i32`. Looking at `-expr as i32` I have no idea whether the `as` or the `-` are executed first, and I might care about that (in particular when casting to an unsigned type).

Comment (KisaragiEffective) [reply src/tools/miri/tests/pass/float.rs:201]: I agreed :)

Comment (RalfJung) [reply src/tools/miri/tests/pass/float.rs:201]: Thanks! The diff in the Miri test can hopefully be removed then. :)

File: tests/ui/lint/unused_parens_cast.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    //~^ WARN unnecessary parentheses around cast expression
+    // following is technically unnecessary, but is allowed because it may confusing.
+    let _ = (-1.0f32) as f64;
+    let x = Box::new(-1.0f32);
+    let _ = (*x) as f64;
+    //~^ WARN unnecessary parentheses around cast expression
+    // cast is left-assoc
+    let _ = (true as u8) as u16;
+    //~^ WARN unnecessary parentheses around cast expression
+    // should not fire
```
Comment (RalfJung) [inline tests/ui/lint/unused_parens_cast.rs:49]: Please add unary `-` to this block (and other unary operator as well?)

Comment (KisaragiEffective) [reply tests/ui/lint/unused_parens_cast.rs:49]: Does your intention include `*expr as Ty` and `&expr as Ty`?

**Comment by Dylan-DPC** [comment]
@KisaragiEffective any updates on it?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**